### PR TITLE
Rework the license header check

### PR DESCRIPTION
**Pull request by slfuchs**
as introduced in #27.

Now three versions are provided:
* `utilities/license-header/run-check-on-current-head.sh` to check all files on current HEAD, e.g., for nightly tests.
* `utilities/license-header/run-check-on-current-branch.sh` to check all files changed on branch with respect to master, e.g., for pull requests
* `utilities/license-header/run-check-on-indexed-files.sh` to check indexed files prior to a commit, e.g., for pre-commit hooks (to be added in the future)

### CODE REVIEW

**Comment by slfuchs** [comment]
I don't know why the GitHub workflow is failing. On my local workstation it just runs perfectly. I'll try to fix that the next days.

File: utilities/license-header/helper-functions.sh [hunk 1]
```diff
@@ -20,7 +20,49 @@
...
+  if [[ "$yearadded" == "$yearlastmod" ]]; then
+    years=$yearadded
+  else
+    years=$yearadded-$yearlastmod
+  fi
+
+  echo $years
+}
+
+exit 0
```
Comment (slfuchs) [inline utilities/license-header/helper-functions.sh:68]: Sorry for the (endless) failing workflows. Now I found out that this `exit 0` here is important when loading this file via `source utilities/license-header/helper-functions.sh`.

**Comment by slfuchs** [comment]
@ppraegla This PR is ready for review! Let's not be super strict with it :wink: I guess when adding pre-commit hooks later on some minor adaptions might be necessary. So no need to over-engineer it now.

File: utilities/license-header/check-license-header.sh [hunk 2]
```diff
@@ -20,39 +20,26 @@
...
-filename=$(basename $filepath)
-
-# examine file type and determine comment symbol
-if [[ $filename == CMakeLists.txt || $filename == *.cmake || $filename == *.sh ]]; then
-  comment="##"
-elif [[ $filename == *.h || $filename == *.h.in || $filename == *.cc ]]; then
-  comment="\/\/" # note: escape the forward slashes
-elif [[ $fileext == *.rst ]]; then
-  comment=".."
+if [[ ! -z $2 ]]; then
```
Comment (ppraegla) [inline utilities/license-header/check-license-header.sh]: The `-n` option also checks for a non empty string.
```suggestion
if [[ -n $2 ]]; then
```

File: utilities/license-header/check-license-header.sh [hunk 3]
```diff
@@ -20,39 +20,26 @@
...
-# get the year the file is added to the repository
-yearadded=$(git log --diff-filter=A --follow --find-renames=90% --format=%ad --date=format:'%Y' -- $filepath | tail -n 1)
-
-# get the year of the last modification of the file
-yearlastmod=$(git log --format="%ad" --date=format:'%Y' --max-count=1 -- $filepath)
-
-# generate range of years for license header
-if [[ "$yearadded" == "$yearlastmod" ]]; then
-  years=$yearadded
+if [[ ! -z $3 ]]; then
```
Comment (ppraegla) [inline utilities/license-header/check-license-header.sh]:
```suggestion
if [[ -n $3 ]]; then
```

File: utilities/license-header/helper-functions.sh [hunk 4]
```diff
@@ -20,7 +20,49 @@
...
+  else
+    comment=""
+  fi
+
+  echo $comment
+}
+
+generate_range_of_years()
+{
+  if [[ ! -z $1 && ! -z $2 ]]; then
```
Comment (ppraegla) [inline utilities/license-header/helper-functions.sh]: Again, I would use `-n` instead of `! -z`

**Review by ppraegla — COMMENTED** [review]
Nice work! Thanks. Only some minor remarks.

With a bit of over-engineering one could probably write another helper function that does the license check as it is quite similar for the three checks, loop over the relevant files and call the helper function. But we can leave that for later ;)

File: utilities/license-header/check-license-header.sh (hunk 2 above)
Comment (slfuchs) [reply utilities/license-header/check-license-header.sh]: Good catch! I adapted that everywhere.

**Comment by slfuchs** [comment]
> With a bit of over-engineering one could probably write another helper function that does the license check as it is quite similar for the three checks, loop over the relevant files and call the helper function. But we can leave that for later ;)

Actually not that easy, as setting the value for `yearadded` and `yearlastmod` in the case of indexed files (that are not yet in the repo) is a different than in the case of files that are already in the repo.

**Review by ppraegla — APPROVED** [review]

**Comment by slfuchs** [comment]
@ppraegla Feel free to merge if you agree with the latest changes.

**Pull request merged**